chipsec/chipset.py

```diff
@@ -912,24 +927,59 @@ def write_register(self, reg_name, reg_value, cpu_thread=0, bus_index=0):
             raise RegisterTypeNotFoundError("Register type not found: {}".format(rtype))
 
     def write_register_all(self, reg_name, reg_values, cpu_thread=0):
-        bus_data = self.get_register_bus( reg_name )
-        if not bus_data:
-            return False
-        values = len(bus_data)
-        if len(reg_values) == values:
-            for index in range( values ):
-                self.write_register( reg_name, reg_values[index], cpu_thread, index )
-            return True
+        reg = self.get_register_def(reg_name)
+        rtype = reg['type']
+        bus_data = self.get_register_bus(reg_name)
+        ret = False
+        if RegisterType.MSR == rtype:
+            topology = self.cpu.get_cpu_topology()
+            if 'scope' in reg.keys() and reg['scope'] == "packages":
+                packages = topology['packages']
+                threads_to_use = [packages[p][0] for p in packages]
+            elif 'scope' in reg.keys() and reg['scope'] == "cores":
+                cores = topology['cores']
+                threads_to_use = [cores[p][0] for p in cores]
+            else: # Default to threads
+                threads_to_use = range(self.helper.get_threads_count())
+            if len(reg_values) == len(threads_to_use):
+                value = 0
+                for t in threads_to_use:
+                    self.write_register(reg_name, reg_values[value], t)
+                    value += 1
+                ret = True
+        elif rtype in [RegisterType.MMCFG, RegisterType.PCICFG, RegisterType.MMIO] and bus_data:
+            values = len(bus_data)
+            if len(reg_values) == values:
```

Comment:
This can become problematic for generalizing tests applicable to single and multi-socket, it means that the user would need to find out what is the current system topology before calling this method, making it less useful. Would it help to make it possible to program a single value to every instance?

Reply:
The expectation would be that the read_register_all is called, and then modify the fields before using the write_register_all command.

Reply:
How about a verbose (HAL?) log if `len(reg_values)` > 1 and != `len(bus_data)`?

Reply:
can you clarify this comment?

Reply:
If `reg_values` doesn't align with `bus_data`, let the user know.  Might help with @iherna2 point about non-aligning topology.

Reply:
Do you want an exception or log?

Reply:
Because the function returns false if the values are not written a log seems appropriate

Reply:
Is this where you were going to add the exception.